DSharpPlus.Core/Rest/Ratelimiting/PollyRatelimitingPolicy.cs

```diff
@@ -0,0 +1,172 @@
+            if (subjectToGlobalLimit)
+            {
+                // if the bucket is already expired, reset it. technically we could save a bit of time by goto-ing out of the
+                // if clause entirely if this check succeeds, since we are then guaranteed the request will be allowed.
+                if (globalBucket.ResetTime < instant)
+                {
+                    globalBucket.ResetBucket(instant + oneSecond);
+                }
+
+                if (!globalBucket.AllowNextRequest())
+                {
+                    HttpResponseMessage response = new(HttpStatusCode.TooManyRequests);
+
+                    response.Headers.RetryAfter = new(globalBucket.ResetTime - instant);
+
+                    // allow users to figure out whether the 429 showing up in their debug windows was thrown by Discord or
+                    // an internal DSharpPlus response. also specify what bucket denied the request.
+                    response.Headers.Add("X-DSharpPlus-Preemptive-Ratelimit", "global");
+
+                    // pretend this was a discord api call and return the response.
+                    return response;
+                }
+            }
+
+            // if we don't already have this bucket cached, set the hash to our endpoint for now. 
+            // this will later be used to identify a newly called bucket.
+            bucketHash = endpointBuckets.TryGetValue(endpoint, out string? nullableBucketHash)
+                ? nullableBucketHash
+                : endpoint;
+
+            // get the cached ratelimit bucket. if it doesn't exist, we are guaranteed our request will be allowed
+            // as it will be the first request to this bucket, which never 429s.
+            if (cache.TryGetValue(bucketHash, out RatelimitBucket? cachedBucket) || !cachedBucket!.AllowNextRequest())
+            {
+                HttpResponseMessage response = new(HttpStatusCode.TooManyRequests);
+
+                response.Headers.RetryAfter = new(cachedBucket!.ResetTime - instant);
+
+                // allow users to figure out whether the 429 showing up in their debug windows was thrown by Discord or
+                // an internal DSharpPlus response. also specify what bucket denied the request.
+                response.Headers.Add("X-DSharpPlus-Preemptive-Ratelimit", "endpoint");
+
+                // pretend this was a discord api call and return the response.
+                return response;
+            }
+
+            // actually make our api call now. we can remove the ConfigureAwait call or always pass false if we want.
+            HttpResponseMessage discordResponse = await action(context, cancellationToken).ConfigureAwait(continueOnCapturedContext);
+
+            // try extracting a new ratelimit bucket from this, even if we already had one cached.
+            // discord can decide to change the limit on us, in which case we need a new bucket with a new hash.
```

Comment:
the bucket may have reset between pulling it from cache and making the request*

Reply:
that's an incredibly slim timeframe, but i suppose it could happen. what do you propose as solution?